File: typescript/agent-calculator.ts
/**
 * LangChain agent with a calculator tool (createOpenAIToolsAgent + AgentExecutor).
 */

import { ChatOpenAI } from "@langchain/openai";
import { ChatPromptTemplate, MessagesPlaceholder } from "@langchain/core/prompts";
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { createOpenAIToolsAgent, AgentExecutor } from "langchain/agents";

const model = new ChatOpenAI({
  model: "gpt-4o",
  temperature: 0,
});

const calculator = tool(
  (input: { expression: string }) => {
    try {
      const expr = input.expression
        .replace(/\bsqrt\s*\(\s*([^)]+)\s*\)/g, "Math.sqrt($1)")
        .replace(/\b(\d+)\s*\/\s*(\d+)/g, "($1/$2)");
      const value = new Function(`return (${expr})`)();
      return String(value);
    } catch {
      return "Error: could not evaluate expression.";
    }
  },
  {
    name: "calculator",
    description: "Evaluate a numeric math expression. Supports +, -, *, /, and sqrt(x).",
    schema: z.object({
      expression: z.string(),
    }),
  }
);

const tools = [calculator];

const prompt = ChatPromptTemplate.fromMessages([
  ["system", "You are a helpful assistant. Use the calculator tool when you need to compute numbers."],
  ["human", "{input}"],
  new MessagesPlaceholder("agent_scratchpad"),
]);

async function runAgent(task: string): Promise<string> {
  const agent = await createOpenAIToolsAgent({
    llm: model,
    tools,
    prompt,
  });
  const executor = new AgentExecutor({ agent, tools });
  const result = await executor.invoke({ input: task });
  return (result.output as string) ?? String(result);
}

async function main(): Promise<void> {
  const task =
    process.argv.slice(2).join(" ") ||
    "Calculate the square root of 144 and then multiply the result by 5.";
  try {
    console.log("Task:", task);
    const result = await runAgent(task);
    console.log("\nResult:", result);
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}

main();

File: java/src/main/java/aibuffet/AgentCalculator.java
package aibuffet;

import dev.langchain4j.agent.tool.Tool;
import dev.langchain4j.model.openai.OpenAiChatModel;
import dev.langchain4j.service.AiServices;

/**
 * LangChain4j agent with a calculator tool (AiServices + @Tool).
 */
public final class AgentCalculator {

    static class CalculatorTools {
        @Tool("Evaluate a numeric math expression. Supports +, -, *, /, and sqrt(x).")
        String calculator(String expression) {
            try {
                String expr = expression
                        .replaceAll("(?i)sqrt\\s*\\(\\s*([^)]+)\\s*\\)", "Math.sqrt($1)")
                        .replaceAll("(\\d+)\\s*/\\s*(\\d+)", "($1/$2)");
                Object value = new javax.script.ScriptEngineManager()
                        .getEngineByName("JavaScript")
                        .eval(expr);
                return String.valueOf(value);
            } catch (Exception err) {
                return "Error: could not evaluate expression.";
            }
        }
    }

    interface Assistant {
        String chat(String userMessage);
    }

    public static void main(String[] args) {
        String task = args.length > 0
                ? String.join(" ", args)
                : "Calculate the square root of 144 and then multiply the result by 5.";

        String apiKey = System.getenv("OPENAI_API_KEY");
        if (apiKey == null || apiKey.isBlank()) {
            System.err.println("OPENAI_API_KEY is not set");
            System.exit(1);
        }

        OpenAiChatModel model = OpenAiChatModel.builder()
                .apiKey(apiKey)
                .modelName("gpt-4o")
                .temperature(0.0)
                .build();

        Assistant assistant = AiServices.builder(Assistant.class)
                .chatLanguageModel(model)
                .tools(new CalculatorTools())
                .build();

        try {
            System.out.println("Task: " + task);
            String result = assistant.chat(task);
            System.out.println("\nResult: " + result);
        } catch (Exception err) {
            System.err.println(err.getMessage());
            System.exit(1);
        }
    }
}
